### ISSUE TITLE
IO latency skyrockets during sequential scrub

### ISSUE THREAD

**Issue by DeHackEd** [zfsonlinux/zfs#6921]
### System information
Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | CentOS
Distribution Version    | 6.9
Linux Kernel                 | 4.9.62 (custom build)
Architecture                 | x86_64
ZFS Version                  | Git d4677269f286005768ae1a0fcd3389aa6015c4c7
SPL Version                  | Git zfsonlinux/spl@ed19bccfb651843fa208232b3a2d3d22a4152bc8

### Describe the problem you're observing
Since sequential scan was added (d4a72f23863382bdf6d0ae33196f5b5decbc48fd) performance of IO has been abysmal during a scrub. Playback of a simple 720p video during a scrub from the pool is not possible. Application latencies are on the order of multiple seconds as shown in the iostats below.

### Describe how to reproduce the problem
Any scrub of a large pool should be sufficient to reproduce the issue.

### Include any warning/errors/backtraces from the system logs
```
# zpool iostat -vl 1
                                                  capacity     operations     bandwidth    total_wait     disk_wait    syncq_wait    asyncq_wait  scrub
pool                                            alloc   free   read  write   read  write   read  write   read  write   read  write   read  write   wait
----------------------------------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
dehacked                                        6.46T  22.5T    148      0  3.68M      0     8s      -  196ms      -      -      -     7s      -      -
  raidz2                                        6.46T  22.5T    147      0  3.65M      0     8s      -  196ms      -      -      -     7s      -      -
    usb-WDC_WD40_EFRX-68N32N0_152D00539000-0:0      -      -     16      0   475K      0     8s      -  195ms      -      -      -     7s      -      -
    usb-WDC_WD40_EFRX-68N32N0_152D00539000-0:1      -      -     18      0   435K      0     8s      -  195ms      -      -      -     7s      -      -
    usb-WDC_WD40_EFRX-68N32N0_152D00539000-0:2      -      -     17      0   427K      0     8s      -  195ms      -      -      -     7s      -      -
    usb-WDC_WD40_EFRX-68N32N0_152D00539000-0:3      -      -     18      0   455K      0     8s      -  201ms      -      -      -     7s      -      -
    usb-ST4000VN_008-2DR166_152D00539000-0:4        -      -     18      0   443K      0     8s      -  195ms      -      -      -     7s      -      -
    usb-ST4000VN_008-2DR166_152D00539000-0:5        -      -     18      0   435K      0     8s      -  195ms      -      -      -     7s      -      -
    usb-ST4000VN_008-2DR166_152D00539000-0:6        -      -     18      0   535K      0     8s      -  195ms      -      -      -     7s      -      -
    usb-ST4000VN_008-2DR166_152D00539000-0:7        -      -     17      0   535K      0     8s      -  201ms      -      -      -     7s      -      -
----------------------------------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
dehacked                                        6.46T  22.5T    143      0  3.16M      0    12s      -  196ms      -      -      -    12s      -      -
  raidz2                                        6.46T  22.5T    143      0  3.16M      0    12s      -  196ms      -      -      -    12s      -      -
    usb-WDC_WD40_EFRX-68N32N0_152D00539000-0:0      -      -     17      0   371K      0    12s      -  195ms      -      -      -    12s      -      -
    usb-WDC_WD40_EFRX-68N32N0_152D00539000-0:1      -      -     17      0   355K      0    12s      -  195ms      -      -      -    12s      -      -
    usb-WDC_WD40_EFRX-68N32N0_152D00539000-0:2      -      -     17      0   487K      0    12s      -  200ms      -      -      -    12s      -      -
    usb-WDC_WD40_EFRX-68N32N0_152D00539000-0:3      -      -     18      0   515K      0    12s      -  190ms      -      -      -    12s      -      -
    usb-ST4000VN_008-2DR166_152D00539000-0:4        -      -     17      0   407K      0    12s      -  200ms      -      -      -    12s      -      -
    usb-ST4000VN_008-2DR166_152D00539000-0:5        -      -     17      0   431K      0    12s      -  195ms      -      -      -    12s      -      -
    usb-ST4000VN_008-2DR166_152D00539000-0:6        -      -     17      0   331K      0    12s      -  195ms      -      -      -    12s      -      -
    usb-ST4000VN_008-2DR166_152D00539000-0:7        -      -     16      0   335K      0    12s      -  194ms      -      -      -    12s      -      -
----------------------------------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----

# iostat -dmx 1 sd[e-l]
Device:         rrqm/s   wrqm/s     r/s     w/s    rMB/s    wMB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sde               0.00     0.00   17.00    0.00     0.37     0.00    44.24     3.00  169.71  169.71    0.00  58.82 100.00
sdf               0.00     0.00   17.00    0.00     0.40     0.00    48.47     3.00  178.00  178.00    0.00  58.82 100.00
sdg               0.00     0.00   16.00    0.00     0.35     0.00    45.00     3.00  172.56  172.56    0.00  62.50 100.00
sdh               0.00     0.00   18.00    0.00     0.45     0.00    51.56     3.00  165.72  165.72    0.00  55.56 100.00
sdi               0.00     0.00   16.00    0.00     0.34     0.00    44.00     3.00  179.69  179.69    0.00  62.50 100.00
sdj               0.00     0.00   18.00    0.00     0.36     0.00    41.33     3.00  163.72  163.72    0.00  55.56 100.00
sdk               0.00     0.00   18.00    0.00     0.42     0.00    47.56     3.00  167.83  167.83    0.00  55.56 100.00
sdl               0.00     0.00   18.00    0.00     0.42     0.00    48.00     3.00  169.22  169.22    0.00  55.56 100.00

Device:         rrqm/s   wrqm/s     r/s     w/s    rMB/s    wMB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sde               0.00     0.00   18.00    0.00     0.50     0.00    57.33     3.00  169.00  169.00    0.00  55.56 100.00
sdf               0.00     0.00   19.00    0.00     0.52     0.00    56.00     3.00  161.32  161.32    0.00  52.63 100.00
sdg               0.00     0.00   19.00    0.00     0.47     0.00    50.95     3.00  167.58  167.58    0.00  52.63 100.00
sdh               0.00     0.00   19.00    0.00     0.46     0.00    49.68     3.00  160.16  160.16    0.00  52.63 100.00
sdi               0.00     0.00   19.00    0.00     0.48     0.00    51.79     3.00  161.00  161.00    0.00  52.63 100.00
sdj               0.00     0.00   19.00    0.00     0.46     0.00    49.68     3.00  161.16  161.16    0.00  52.63 100.00
sdk               0.00     0.00   19.00    0.00     0.48     0.00    51.79     3.00  158.11  158.11    0.00  52.63 100.00
sdl               0.00     0.00   18.00    0.00     0.48     0.00    55.11     3.00  163.00  163.00    0.00  55.56 100.00

  pool: dehacked
 state: ONLINE
status: Some supported features are not enabled on the pool. The pool can
	still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
	the pool may no longer be accessible by software that does not support
	the features. See zpool-features(5) for details.
  scan: scrub paused since Sun Dec  3 10:44:30 2017
	scrub started on Sun Dec  3 10:44:30 2017
	4.53T scanned, 576K issued, 6.46T total
	0B repaired, 0.00% done
config:

	NAME                                            STATE     READ WRITE CKSUM
	dehacked                                        ONLINE       0     0     0
	  raidz2-0                                      ONLINE       0     0     0
	    usb-WDC_WD40_EFRX-68N32N0_152D00539000-0:0  ONLINE       0     0     0
	    usb-WDC_WD40_EFRX-68N32N0_152D00539000-0:1  ONLINE       0     0     0
	    usb-WDC_WD40_EFRX-68N32N0_152D00539000-0:2  ONLINE       0     0     0
	    usb-WDC_WD40_EFRX-68N32N0_152D00539000-0:3  ONLINE       0     0     0
	    usb-ST4000VN_008-2DR166_152D00539000-0:4    ONLINE       0     0     0
	    usb-ST4000VN_008-2DR166_152D00539000-0:5    ONLINE       0     0     0
	    usb-ST4000VN_008-2DR166_152D00539000-0:6    ONLINE       0     0     0
	    usb-ST4000VN_008-2DR166_152D00539000-0:7    ONLINE       0     0     0

```

### Other
```
options zfs zfs_arc_max=9000000000 l2arc_norw=0
options spl spl_taskq_thread_dynamic=0
```

Pool is made of 8 disks in a single raidz2 on a USB 3.0 8-drive external enclosure. Latency is a bit high overall as a result but this is clearly ZFS not issuing IO very well.

**Comment by behlendorf** [comment]
@DeHackEd indeed, from the stats you posted it clearly looks like we're starving asynchronous reads when scrubbing.  That's by design since we want low priority predictive prefetches to opportunistically fill in any gaps.  The problem here is very likely the one described in [arc_read()](https://github.com/zfsonlinux/zfs/blob/master/module/zfs/arc.c#L5915), specifically there is no existing mechanism to convert a pending prefetch in to a demand read which would be serviced immediately.  It sounds like we're going to need to implement this to resolve this performance issue.

As an immediate workaround try setting `zfs_scan_legacy=1` to revert to the previous scrubbing algorithm.

**Comment by tcaputi** [comment]
I agree with @behlendorf 's assessment. @DeHackEd to be sure that this is correct, can you post the output of `cat /proc/spl/kstat/zfs/arcstats | grep sync_wait_for_async`? This will tell us how many times we wanted to change an IO's priority and weren't able to. I think you can also prevent this issue by setting `zfs_prefetch_disable=1`. This should disable predictive prefetch altogether, while still allowing for sequential scrubs. That's not a long-term fix, but I think it will get you back up and running.

**Comment by tcaputi** [comment]
If you get the chance, could you try this diff and see if it helps? If this is enough to solve the problem, its a pretty easy fix: https://pastebin.com/5BG2QAUS

Sorry, my large testing pool is in use.

**Comment by DeHackEd** [comment]
Can confirm `sync_wait_for_async` is climbing slowly (many seconds between increments). But that also describes video playback from the pool during this scrub.

I'll apply that commit as soon as I can. Right now I don't want to shut down my current pools either, but I should be able to apply it in a few days.

**Comment by behlendorf** [comment]
@tcaputi can you open a PR with your patch.

**Comment by tcaputi** [comment]
Sure. I was going to wait until we knew if it worked, but I can do it now.

**Comment by grwilson** [comment]
@DeHackEd can you try to dynamically increase `zfs_vdev_async_read_max_active` to see if it helps with the playback performance?

**Comment by DeHackEd** [comment]
Subjectively it was a bit better, but cranking it up to 10 or even 15 wasn't enough to improve it to what would be considered acceptable for playback.

Of note is that performance improves dramatically once you hit the sequential dispatch stage.

**Comment by tcaputi** [comment]
@dehacked is that with the tunable set or with the new code?

**Comment by DeHackEd** [comment]
Both
